libcloud/storage/drivers/runabove.py

```diff
@@ -0,0 +1,289 @@
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from libcloud.utils.py3 import httplib
+from libcloud.common.base import ConnectionKey
+from libcloud.common.types import ProviderError
+from libcloud.storage.base import Object, Container, StorageDriver
+from libcloud.storage.types import ContainerIsNotEmptyError
+from libcloud.storage.types import ContainerDoesNotExistError
+from libcloud.storage.types import ObjectDoesNotExistError
+from libcloud.common.runabove import API_ROOT, RunAboveConnection
+from libcloud.common.runabove import RunAboveResponse
+
+
+class ContainerConnection(ConnectionKey):
+    responseCls = RunAboveResponse
+
+    def add_default_headers(self, headers):
+        headers.update({
+            'X-Auth-Token': self.key
+        })
+        return headers
+
+
+class RunAboveContainer(Container):
+    def get_object(self, object_name):
+        return self.driver.get_object(container_name=self.name,
+                                      object_name=object_name,
+                                      ex_location=self.extra['region'])
+
+
+class RunAboveStorageDriver(StorageDriver):
+    """
+    Libcloud driver for the RunAbove API
+
+    For more information on the RunAbove API, read the official reference:
+
+        https://api.runabove.com/console/
+    """
+    name = 'RunAbove'
+    website = 'https://www.runabove.com/'
+    connectionCls = RunAboveConnection
+    containerConnectionCls = ContainerConnection
+    hash_type = 'md5'
+    supports_chunked_encoding = False
+    ex_location = None
+
+    def __init__(self, key, secret, ex_consumer_key=None, ex_location=None):
+        """
+        Instantiate the driver with the given API credentials.
+
+        :param key: Your application key (required)
+        :type key: ``str``
+
+        :param secret: Your application secret (required)
+        :type secret: ``str``
+
+        :param ex_consumer_key: Your consumer key (required)
+        :type ex_consumer_key: ``str``
+
+        :rtype: ``None``
+        """
+        self.datacenter = None
+        self.consumer_key = ex_consumer_key
+        self.location = ex_location
+        StorageDriver.__init__(self, key, secret,
+                               ex_consumer_key=ex_consumer_key)
+        response = self.connection.request(action='%s/token' % API_ROOT)\
+            .object
+        self.auth_token = response["X-Auth-Token"]
```

Comment:
Not a big fan of constructors with side-affects.

Is it possible to move this to the connection class and perform it lazily on the first request?